modal-js/src/sandbox_filehandle.ts

```diff
@@ -0,0 +1,154 @@
+import {
+  ContainerFilesystemExecRequest,
+  DeepPartial,
+  ContainerFilesystemExecResponse,
+} from "../proto/modal_proto/api";
+import { client, isRetryableGrpc } from "./client";
+import { RemoteError } from "./errors";
+
+/** File open modes supported by the filesystem API. */
+export type SandboxFileMode = "r" | "w" | "a" | "r+" | "w+" | "a+";
+
+export type SandboxReadOptions = {
+  /** Encoding for text operations. Defaults to 'utf8' for utf8 encoded data. */
+  encoding?: "utf8" | "binary";
+};
+
+/**
+ * SandboxFileHandle represents an open file in the sandbox filesystem.
+ * Provides read/write operations similar to Node.js `fsPromises.FileHandle`.
+ */
+export class SandboxFileHandle {
+  readonly #fileDescriptor: string;
+  readonly #taskId: string;
+
+  /** @ignore */
+  constructor(fileDescriptor: string, taskId: string) {
+    this.#fileDescriptor = fileDescriptor;
+    this.#taskId = taskId;
+  }
+
+  /**
+   * Read data from the file.
+   * @returns Promise that resolves to the read data as Uint8Array or string
+   */
+  async read(): Promise<Uint8Array>;
+  async read(
+    options: SandboxReadOptions & { encoding: "utf8" },
+  ): Promise<string>;
+  async read(
+    options: SandboxReadOptions & { encoding: "binary" },
+  ): Promise<Uint8Array>;
+  async read(options?: SandboxReadOptions): Promise<Uint8Array | string> {
```

Comment:
I think I'd want to think more about proper Unicode text support before we merge `encoding: "utf8"` support, since it's easy to end up with corrupted data. Just to get this PR in, can we stick to binary format? Happy to prioritize thinking through this API as a follow-up if it's a big ergonomic hassle for users.